flfm/app/tests/test_app.py

```diff
@@ -0,0 +1,224 @@
+import base64
+import time
+from copy import deepcopy
+from multiprocessing import Process
+
+import numpy as np
+import pytest
+import requests
+from dash import dcc, html
+from plotly.graph_objs import Figure
+
+import flfm.app.main
+import flfm.io
+import flfm.util
+from flfm.settings import app_settings, settings
+
+DATA_DIR = flfm.util.find_package_location() / "tests" / "data" / "yale"
+PSF_FILENAME = DATA_DIR / "measured_psf.tif"
+IMAGE_FILENAME = DATA_DIR / "light_field_image.tif"
+IDS = ("psf", "image", "reconstruction")
+FILENAMES = {IDS[0]: PSF_FILENAME, IDS[1]: IMAGE_FILENAME}
+
+
+def start_real_server():
+    # Start a real server on a separate process. It's easier to kill a process than a thread.
+    # For the majority of tests we can just use ``dash.testing`` or ``fastapi.testclient`` instead of this. However, it
+    # would be good to still test prd deployment run code.
+
+    proc = Process(target=flfm.app.main.start_app)
+    proc.start()
+    time.sleep(5)  # Both the new process & server take time to start up.
+
+    if not proc.is_alive():
+        # raise RuntimeError("Server did not start")
+        ...
+
+    def _kill():
+        proc.kill()
+        proc.join()
+        return proc
+
+    return proc, _kill
+
+
+class TestUtility:
+    def test_startup_tasks(self, monkeypatch):
+        monkeypatch.setattr(app_settings, "RUN_DUMMY_RECONSTRUCTION_AT_STARTUP", True)
+        monkeypatch.setattr(app_settings, "DUMMY_PSF_FILEPATH", PSF_FILENAME)
+        monkeypatch.setattr(app_settings, "DUMMY_LIGHT_FILED_IMAGE_FILEPATH", IMAGE_FILENAME)
+
+        flfm.app.main.startup_tasks()
+
+    def test_center_frame(self):
+        assert flfm.app.main._center_frame(40) == 20
+
+    def test_slider_template(self):
+        assert flfm.app.main.slider_template_str(40, 20, 5) == " 100.00 µm"
+
+    @pytest.mark.parametrize("debug", (True, False))
+    def test_exception_handler_no_exceptions(self, monkeypatch, debug):
+        monkeypatch.setattr(app_settings, "DEBUG", debug)
+
+        error = RuntimeError()
+        if debug:
+            with pytest.raises(type(error)):
+                flfm.app.main.exception_handler(error)
+        else:
+            flfm.app.main.exception_handler(error)
+
+
+class TestRunServer:
+    def test_wrong_api(self, monkeypatch):
+        monkeypatch.setattr(app_settings, "WEB_API", "blah")
+        with pytest.raises(NotImplementedError):
+            flfm.app.main.start_app()
+
+    def test_flask_debug_mode_exception(self, monkeypatch):
+        monkeypatch.setattr(app_settings, "WEB_API", "flask")
+        monkeypatch.setattr(app_settings, "DEBUG", False)
+
+        with pytest.raises(NotImplementedError):
+            flfm.app.main.start_app()
+
+    @pytest.mark.parametrize("api", ("flask", "fastapi"))
+    def test_run_server(self, api, monkeypatch):
+        if api == "flask":
+            pytest.skip("Broken from pytest. See https://github.com/ssec-jhu/flfm/issues/174")
```

Comment:
@ryanhausen FYI I couldn't get this to work from pytest so I have punted to #174 rather than have it block this PR.